webnn/resources/utils.js

```diff
@@ -208,86 +230,159 @@ const assert_array_approx_equals_ulp = (actual, expected, nulp, dataType, descri
+    assert_array_approx_equals(actual, expected, tolerance, description);
+  }
+};
+
+/**
+ * Check computed results with expected data.
+ * @param {String} name - An operation name
  * @param {Object.<String, MLOperand>} namedOutputOperands
  * @param {Object.<MLNamedArrayBufferViews>} outputs - The resources of required outputs
  * @param {Object} resources - Resources used for building a graph
  */
-const checkResults = (opName, operandType, namedOutputOperands, outputs, resources) => {
-  const metricType = PrecisionMetrics[opName] ? Object.keys(PrecisionMetrics[opName])[0] : 'ULP';
-  const description = `test ${opName} ${operandType}`;
-  const tolerance = getPrecisonTolerance(opName, metricType, operandType, resources);
-  if (metricType === 'ULP') {
+const checkResults = (name, namedOutputOperands, outputs, resources) => {
+  const operandType = resources.type;
+  const metricType = PrecisionMetrics[name] ? Object.keys(PrecisionMetrics[name])[0] : 'ULP';
+  const tolerance = getPrecisonTolerance(name, metricType, resources);
+  const expected = resources.expected;
+  let outputData;
+  let expectedData;
+  if (Array.isArray(expected)) {
+    // check outputs by split or gru
     for (let operandName in namedOutputOperands) {
-      let outputData = outputs[operandName];
-      let expectedData = resources.expected.data[operandName];
-      if (typeof expectedData === 'number') {
-        outputData = [outputData];
-        expectedData = [expectedData];
-      }
-      assert_array_approx_equals_ulp(outputData, expectedData, tolerance, operandType, description);
+      outputData = outputs[operandName];
+      expectedData = getExpectedData(expected, operandName);
+      doAssert(name, outputData, expectedData, tolerance, operandType, metricType)
     }
-  } else if (metricType === 'ATOL') {
-    for (let operandName in namedOutputOperands) {
-      assert_array_approx_equals(
-        outputs[operandName], resources.expected.data[operandName], tolerance, description);
+  } else {
+    const outputName = expected.name ? expected.name : 'output';
+    outputData = outputs[outputName];
+    expectedData = expected.data;
+    doAssert(name, outputData, expectedData, tolerance, operandType, metricType)
+  }
+};
+
+/**
+ * Create input operands for a graph.
+ * @param {MLGraphBuilder} builder - A ML graph builder
+ * @param {Object} resources - Resources used for building a graph
+ * @param {String[]} nameArray
+ * @returns {MLOperand[]} Input operands
+ */
+const createInputOperands = (builder, resources, nameArray) => {
+  let operands = [];
+  nameArray.forEach(name => {
+    operands.push(builder.input(name, {type: resources.type, dimensions: resources.inputs[name].shape}));
+  });
+  return operands;
+};
+
+/**
+ * Build a graph.
+ * @param {MLGraphBuilder} builder - A ML graph builder
+ * @param {Object} resources - Resources used for building a graph
+ * @param {Function} computeFunc - A compute function
```

Comment:
The name is confusing. It is actual a build function that build a specific operation.

Reply:
Thanks, I will update it.